libkbfs/prefetcher.go

```diff
@@ -612,6 +725,16 @@ func (p *blockPrefetcher) run(testSyncCh <-chan struct{}) {
 				p.log.CDebugf(ctx, "created new prefetch for block %s",
 					req.ptr.ID)
 			}
+
+			if p.rescheduleIfNeeded(ctx, req) {
```

Comment:
@jzila we can't actually put this call here, because of these lines above: https://github.com/keybase/kbfs/blob/0d0a1c8f5c24918d8c1fa745efd2c9ababecde44/libkbfs/prefetcher.go#L713-L715

The block could have already been decremented and marked as subtree-triggered, but then when we reschedule it here we don't undo any of that, and some prefetches will never be completed due to the incorrect state.

So I'm inclined to move it back up to where it was.  Or I could try to refactor this code a bit so that the single `continue` above happens, then we do this check, and then we do everything else that might modify the prefetch request state.  Let me know if you have a preference.

Reply:
If we have `a -> b -> c`, and we are handling `c`, then we'll:
* decrement up to `a`
* set `subtreeTriggered`
* reschedule `c`
  * lose `subtreeTriggered` (which we actually want, since it hasn't happened until `handlePrefetch`)
  * cancel `c` and `b`, rescheduling `a`.

Seems to me like we do actually want to lose this information. What seems to be going wrong?

Reply:
@jzila Here's the exact situation:

1. Path `root -> {a -> {aa, ab}, b}`.
2. Cache fills up after `a` and `ab` are fetched.
3. When handling the request for `aa`, the parent path is decremented, bring `a` down to a count of 0.  But `completePrefetch` is never called, so `a` never gets the status `FinishedPrefetch`.
4. Then `aa` is queued to be rescheduled.
5. The test explicitly requests `a`.
6. The prefetcher sees that `a` has a `subtreeBlockCount` of 0, and cancel its prefetch request.
7. When `aa` is rescheduled, its parent doesn't have an entry in `p.prefetches` because it was just canceled, and so `aa` is seen as the top-block to be rescheduled.
8. Once `aa` is fetched, it completes itself, but never completes `a` since it's no longer listed as a parent.

So neither `a` nor `root` are ever marked as `FinishedPrefetch`.  And it happens because the block counts were decremented before a reschedule.

Reply:
I see, okay. We can move it back and leave the slight efficiency losses alone.